package/MDAnalysis/guesser/base.py

```diff
@@ -0,0 +1,116 @@
+# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding: utf-8 -*-
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
+#
+# MDAnalysis --- https://www.mdanalysis.org
+# Copyright (c) 2006-2017 The MDAnalysis Development Team and contributors
+# (see the file AUTHORS for the full list of names)
+#
+# Released under the GNU Public Licence, v2 or any higher version
+#
+# Please cite your use of MDAnalysis in published work:
+#
+# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
+# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+# doi: 10.25080/majora-629e541a-00e
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+from .. import _GUESSERS
+
+class GuesserMeta(type):
+    def __init__(cls, name, bases, classdict):
+        type.__init__(type, name, bases, classdict)
+        try:
+            _GUESSERS[classdict['context'].upper()] = cls
+        except KeyError:
+            pass
+
+class GuesserBase(metaclass=GuesserMeta):
+    context = 'base' 
+    _guess = {}
+
+    # give a rank to each atrribute based on its dependency on other attributes
+    # to be guessed, so that the attribute with lesser dependcy will be guessed
+    # first
+    _rank = {}
```

Comment:
Having `_guess` and `_rank` being separate attributes is dangerous. Indeed, it is very easy to have one key in one dictionary missing from the other one. It would be safer to have `_guess` being the form `{key: (method, rank)}`. We can think about having a decorator to populate `_guess`:

```python
class MyGuesser(GuesserBase):
    @guesses('elements')  # rank is 0 by default
    def guess_elements(self, ...):
        ...

    @guesses('bonds', rank=1):
    def guess_bonds(self, ...):
        ...
```

Another issue is that it is easy to have the rank wrong. Imagine a case where we have 4 methods to guess attributes: A, B, C, and D. B depends on A, C depends on A as well, and C depends on B and C. So A is rank 0, B and C are rank 1, and D is rank 2. Now we change the code of C so that it depends on both A and B. We update the rank of C, however the contributor that changes C does now know about D so it does not change its rank. now C and D have the same rank so, possibly, guessing D can be attempted before guessing C causing an error.

A more robust solution would be to handle the dependencies directly in the guessing code. Guessing the bonds requires knowing about the elements, so we could have a system so that when the method that guesses the bonds fetches the elements, they get guessed if they are not available. Later on, when we call the method to guess the elements, nothing happens because they are already available.